src/fixate/core/switching.py

```diff
@@ -0,0 +1,743 @@
+        :return:
+        """
+        raise NotImplementedError
+
+    # used in a few scripts
+    def update_pin_by_name(
+        self, name: Pin, value: bool, trigger_update: bool = True
+    ) -> None:
+        raise NotImplementedError
+
+    # not used in any scripts
+    def update_pins_by_name(
+        self, pins: Collection[Pin], trigger_update: bool = True
+    ) -> None:
+        raise NotImplementedError
+
+    def __getitem__(self, item: Pin) -> bool:
+        """Get the value of a pin. (only inputs? or state of outputs also?)"""
+        raise NotImplementedError
+
+    def __setitem__(self, key: Pin, value: bool) -> None:
+        """Set a pin"""
+        raise NotImplementedError
+
+
+class MuxGroup:
+    """
+    Group multiple VirtualMux's, for use in a single Jig Driver.
+
+    If a test script, it is expected that MuxGroup will be subclassed, with attributes
+    being each required VirtualMux subclass. This can be done using a dataclass:
+
+    @dataclass
+    class JigMuxGroup(MuxGroup):
+        mux_one: MuxOne = field(default_factory=MuxOne)
+        mux_two: MuxTwo = field(default_factory=MuxTwo)
+    """
+
+    def get_multiplexers(self) -> list[VirtualMux]:
+        return [attr for attr in self.__dict__.values() if isinstance(attr, VirtualMux)]
+
+    def reset(self) -> None:
+        mux_list = self.get_multiplexers()
+        if len(mux_list) == 0:
+            return
+
+        for mux in mux_list[:-1]:
+            mux.reset(trigger_update=False)
+        mux_list[-1].reset(trigger_update=True)
+
+    def active_signals(self) -> list[str]:
+        return [str(mux) for mux in self.get_multiplexers()]
+
+
+JigSpecificMuxGroup = TypeVar("JigSpecificMuxGroup", bound=MuxGroup)
+
+
+class JigDriver(Generic[JigSpecificMuxGroup]):
+    """
+    Combine multiple VirtualMux's and multiple AddressHandler's.
+
+    The jig driver joins muxes to handlers by matching up pin definitions.
+    """
+
+    def __init__(
+        self,
+        mux_group_factory: Callable[[], JigSpecificMuxGroup],
+        handlers: Sequence[AddressHandler],
+    ):
+        # keep a reference to handlers so that we can close them if required.
+        self._handlers = handlers
+        self.virtual_map = VirtualAddressMap(handlers)
+
+        self.mux = mux_group_factory()
+        for mux in self.mux.get_multiplexers():
+            # Perhaps we should instantiate the virtual mux here
+            # and pass in the virtual_map.add_update. But we'd have to do some
+            # magic in the MuxGroup call to pass add_update to each VirtualMux
+            # constructor, and I was hoping to just use a dataclass...
+            mux._update_pins = self.virtual_map.add_update
+
+    def __setitem__(self, key: Pin, value: bool) -> None:
+        self.virtual_map.update_pin_by_name(key, value)
+
+    def __getitem__(self, item: Pin) -> bool:
+        return self.virtual_map[item]
```

Comment:
I've always hated these due to the lack of clarity of was actually going on and that it seemed like you could bypass everything handled by the virtualmux. I think this implementation is at least a bit nicer now this are a bit less tangled. Would rather an explicit set/get pin if we really want manual control over pins but this should be pretty rare outside of debugging jigs.

Reply:
It is 100% for debugging. Perhaps we should attempt to document than.

Perhaps we should remove these from `JigDriver` and only have them on `jig.virtual_map`. I don't have strong opinions on if we keep the `__getitem__` and `__setitem__` methods, but I think the general functionality is important to keep.

Reply:
I think we should definitely make it clear that these should only be used for debugging.
Because it wouldn't be normal to see something like `jig["pin_name"] = True` I'm not against having the `__getitem__` and `__setitem__` methods.

Maybe prefix methods that are meant only for debugging with debug_ ? Just a thought, but at least it would be very clear that it's for debug only.

Reply:
I'm leaning toward removing the `__setitem__` and `__getitem__` and adding `jig.debug_set_pin(pin, value)`, `jig.debug_set_pins(on: Collection[Pin], off: Collection[Pin])`

Reply:
I support that